### ISSUE TITLE
Add `MonadFail` instance to `Gen` data type

### ISSUE THREAD

**Issue by ChShersh** [hedgehogqa/haskell-hedgehog#225]
Without this instance it's impossible to perform partial pattern-matching inside property-based tests. Maybe this instance can be somehow smart, like, discard all cases that doesn't match the case?

**Comment by ChShersh** [comment]
@moodmosaic I'm actually okay with having `error` for now, works for me as well :+1:

**Comment by ocharles** [comment]
I think discarding is the correct thing to do, personally. One could imagine doing something like

```haskell
Just foo <- forAll ...
```

And having tests run, rather than the test crash.

**Comment by moodmosaic** [comment]
Yes, discarding looks, but also requires a major version bump, because it's a breaking change. In reality though, it's more or less what @sol tries to explain [here](https://github.com/nick8325/quickcheck/pull/228#issuecomment-427535945):

>MonadFail is still fairly new and we as a community don't have the collective experience yet on how things will pan out in practice.

>That you get compile-time errors instead of runtime errors in some situations now is an intentional change. It's the whole point of introducing MonadFail.

I'm really curious to see what happens with https://github.com/nick8325/quickcheck/pull/228.

**Comment by fosskers** [comment]
Any more thoughts here? Or should we just consider it incorrect to pattern match in this way in tests?

**Comment by moodmosaic** [comment]
Follow https://github.com/hedgehogqa/haskell-hedgehog/pull/257 for updates.

**Comment by fosskers** [comment]
Thanks!
